### PR TITLE
ORTOptimizer refactor

### PR DESCRIPTION
**Pull request by michaelbenayoun**
# What does this PR do?

- Renames the attributes for the `OptimizationConfig`: 
    - The previous naming was not very clear, because it was matching the naming from what was done in `onnxruntime` for instance `disable_attention` for disabling attention fusion.
    - Make things clearer in terms of optimization tool by explicitly specifying which attributes tune the general purpose ONNX Runtime optimizer (compared to the other which are tuning the transformers specific optimizer).
- Adds an `AutoOptimizationConfig` class, with several optimization levels that cover the most common use cases, to make it easy to create an `OptimizationConfig` without having to dive in the ONNX Runtime ecosystem. 
- Uses `NormalizedConfig` when possible.

To be done:

- [ ] Write tests for configuration classes

### CODE REVIEW

File: optimum/onnxruntime/configuration.py [hunk 1]
```diff
@@ -604,73 +608,222 @@ def tensorrt(
...
     OptimizationConfig is the configuration class handling all the ONNX Runtime optimization parameters.
+    There are two stacks of optimizations:
+        1. The ONNX Runtime general-purpose optimization tool: it can work on any ONNX model.
+        2. The ONNX Runtime transformers optimization tool: it can only work on a subset of transformers models.
 
-    Args:
-        optimization_level (`int`, defaults to 1):
-            ONNX opset version to export the model with.
+    Attributes:
+        general_optimization_level (`int`, defaults to 1):
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/configuration.py:616]: Not a big fan of this name, but basically trying to convey that this `optimization_level` is about the general ONNX Runtime tool, and not the `transformers`-specific optimizations.

Do you have any better suggestion @echarlaix @regisss ?

File: optimum/onnxruntime/configuration.py [hunk 2]
```diff
@@ -604,73 +608,222 @@ def tensorrt(
...
+                - 99: will enable all available optimizations including layout optimizations
         optimize_for_gpu (`bool`, defaults to `False`):
             Whether to optimize the model for GPU inference.
             The optimized graph might contain operators for GPU or CPU only when `optimization_level` > 1.
         fp16 (`bool`, defaults to `False`):
             Whether all weights and nodes should be converted from float32 to float16.
-        optimize_with_onnxruntime_only (`bool`, defaults to `False`):
-            Whether to only use ONNX Runtime to optimize the model and no graph fusion in Python.
-        disable_gelu (`bool`, defaults to `False`):
+        enable_transformers_specific_optimizations (`bool`, defaults to `True`):
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/configuration.py:629]: I went for `enable_transformers_specific_optimizations`, which I think is nice because it conveys that we will use a `transformers` specific tool.
 @echarlaix @regisss (did not use `fuse_operators` since you also have approximations such as fast gelu)

File: docs/source/onnxruntime/usage_guides/optimization.mdx [hunk 3]
```diff
@@ -108,3 +108,27 @@ Below you will find an easy end-to-end example on how to optimize a Seq2Seq mode
 >>> outputs = optimized_model.generate(**tokens)
 ```
 
+
+## Optimization Configuration
+
+# TODO: edit this before merging, once the names have been chosen.
```
Comment (michaelbenayoun) [inline docs/source/onnxruntime/usage_guides/optimization.mdx:114]: I will edit that once we settled with the names.

File: optimum/onnxruntime/configuration.py (hunk 1 above)
Comment (regisss) [reply optimum/onnxruntime/configuration.py:616]: I think we should keep `optimization_level` here because I have the feeling more accurate names will confuse users. This is not great but I don't think there is a perfect solution.
That means that we need to make it very clear in the documentation.

File: optimum/onnxruntime/configuration.py (hunk 2 above)
Comment (regisss) [reply optimum/onnxruntime/configuration.py:629]: Sounds good to me :+1:

File: optimum/onnxruntime/configuration.py [hunk 4]
```diff
@@ -604,73 +608,222 @@ def tensorrt(
...
+                    f"{old_name} will be deprecated soon, use {new_name} instead.",
+                    FutureWarning,
+                )
+                if mapping_func is None:
+
+                    def identity(x):
+                        return x
+
+                    mapping_func = identity
+                setattr(self, new_name, mapping_func(getattr(self, old_name)))
```
Comment (regisss) [inline optimum/onnxruntime/configuration.py:707]: Maybe we could put the deprecation warning after setting the value of the new attribute and add to the warning message something like: `f"{new_name} is set to {mapping_func(getattr(self, old_name))}"`.

File: optimum/onnxruntime/configuration.py [hunk 5]
```diff
@@ -730,14 +732,87 @@ def create_fusion_options(self, model_type: str) -> FusionOptions:
...
             if hasattr(args, attr):
                 continue
             setattr(args, attr, value)
 
         return FusionOptions.parse(args)
 
 
+class AutoOptimizationConfig:
+    _LEVELS = [
+        {
```
[the remark this answers is not in the document]
Comment (regisss) [reply optimum/onnxruntime/configuration.py:752]: I feel like 6 levels is a lot. Here is a proposal:
- 0: `general_optimization_level=1` and that's it (basic),
- 1: `general_optimization_level=2`, `enable_transformers_specific_optimizations=True` and `enable_gelu_approximation=True` (medium),
- 2: `general_optimization_level=99`, `enable_transformers_specific_optimizations=True`, `enable_gelu_approximation=True` and `fp16=True` (advanced).

I would also maybe name them O0, O1 and O2 to avoid any confusion with ORT optimization levels. WDYT?

File: optimum/onnxruntime/optimization.py [hunk 6]
```diff
@@ -75,29 +85,28 @@ def from_pretrained(cls, model_or_path: Union[str, os.PathLike, ORTModel], file_
...
         else:
             raise ValueError(f"Unable to load the model from {model_or_path}.")
+        return cls(onnx_model_path, config=config)
 
     def optimize(
         self,
         optimization_config: OptimizationConfig,
         save_dir: Union[str, os.PathLike],
-        file_suffix: Optional[str] = "optimized",
+        file_suffix: str = "optimized",
```
Comment (regisss) [inline optimum/onnxruntime/optimization.py:105]: The `Optional` hint should still be there

**Review by regisss — COMMENTED** [review]
It looks good @michaelbenayoun! I left a few additional comments but since it's still a draft feel free to ignore them.

File: optimum/onnxruntime/configuration.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/onnxruntime/configuration.py:752]: I agree!
What is the expected drop in accuracy with `enable_gelu_approximation=True`? If it's significant, I would suggest adding one level with only `general_optimization_level=2, enable_transformers_specific_optimizations=True `

File: docs/source/onnxruntime/usage_guides/optimization.mdx [hunk 7]
```diff
@@ -108,3 +108,27 @@ Below you will find an easy end-to-end example on how to optimize a Seq2Seq mode
...
+In the optimization configuration, there are 4 possible optimization levels:
+- `general_optimization_level=0` to disable all optimizations,
+- `general_optimization_level=1` to enable basic optimizations such as constant folding or redundant node eliminations,
+- `general_optimization_level=2` to enable extended graph optimizations such as node fusions,
+- `general_optimization_level=99` to enable data layout optimizations.
+
+Choosing a level enables the optimizations of that level, as well as the optimizations of all preceding levels.
+More information [here](https://onnxruntime.ai/docs/performance/graph-optimizations.html).
+
+`fuse_operators=True` means that graph fusion is performed in addition to ONNX Runtime optimizations described above.
```
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply docs/source/onnxruntime/usage_guides/optimization.mdx:125]: Also, I will edit that now that we have settled the naming.

File: optimum/exporters/onnx/__main__.py [hunk 8]
```diff
@@ -64,11 +64,11 @@ def main():
         ),
     )
     parser.add_argument("--cache_dir", type=str, default=None, help="Path indicating where to store cache.")
-    parser.add_argument("output", type=Path, help="Path indicating where to store generated ONNX model.")
+    parser.add_argument("output", type=Path, help="Path indicating the directory where to store generated ONNX model.")
 
     # Retrieve CLI arguments
     args = parser.parse_args()
-    args.output = args.output if args.output.is_file() else args.output.joinpath("model.onnx")
+    args.output = args.output.joinpath("model.onnx")
```
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/exporters/onnx/__main__.py:71]: No, but if the file did not exist it was failing, and I feel like it should be consistent => producing a directory with the file inside.
You think we should keep things as they were before?

File: optimum/onnxruntime/modeling_seq2seq.py [hunk 9]
```diff
@@ -711,18 +712,18 @@ def prepare_output_buffer(
...
             output_buffer = torch.empty(1, dtype=torch_type, device=self._device).contiguous()
         elif output_name == "logits":
-            output_shape = (batch_size, sequence_length, self.config.vocab_size)
+            output_shape = (batch_size, sequence_length, self.normalized_config.vocab_size)
             output_buffer = torch.empty(np.prod(output_shape), dtype=torch_type, device=self._device).contiguous()
         elif "key_values" in output_name:
-            num_heads = getattr(self.config, ORTConfigManager.get_num_heads_name(self.config.model_type))
-            hidden_size = getattr(self.config, ORTConfigManager.get_hidden_size_name(self.config.model_type))
-            embed_size_per_head = hidden_size // num_heads
+            num_attention_heads = self.normalized_confi.num_attention_heads
```
Comment (echarlaix) [inline optimum/onnxruntime/modeling_seq2seq.py:718]:
```suggestion
            num_attention_heads = self.normalized_config.num_attention_heads
```

File: optimum/onnxruntime/configuration.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (echarlaix) [reply optimum/onnxruntime/configuration.py:752]: An `AutoOptimizationConfig` could be interesting in the case where we want to optimize a model while keeping it compatible with quantization or if already have a quantized model and want to automatically define the optimizations that can be applied on it. I think we should wait before adding it as the different options could vary as there is no clear options which currently makes more sense than the others to me

**Review by echarlaix — COMMENTED** [review]
It looks great ! 🔥

File: optimum/onnxruntime/configuration.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/onnxruntime/configuration.py]: The current `OptimizationConfig` is not intuitive, hence the `AutoOptimizationConfig` hiding those details to the user. 
Also, we can add those features later, for now the `AutoOptimizationConfig` provides "good" common optimizations, independent from quantization, exactly as what `AutoQuantizationConfig` does for quantization.

File: optimum/exporters/onnx/__main__.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (regisss) [reply optimum/exporters/onnx/__main__.py:71]: No it's fine like this. I was just curious, it sounds good to me :+1:

File: optimum/onnxruntime/configuration.py [hunk 10]
```diff
@@ -604,73 +608,248 @@ def tensorrt(
...
+                Whether the model to optimize will run on GPU, some optimizations depends on the hardware the model
+                will run on. Only needed for optimization_level > 1.
+            kwargs (`Dict[str, Any]`, *optional*):
+                Arguments to provide to the [`~OptimizationConfig`] constructor.
+
+        Returns:
+            `OptimizationConfig`: The `OptimizationConfig` corresponding to the requested optimization level.
+        """
+        if optimization_level not in cls._LEVELS:
+            raise ValueError(f"optimization_level must be in {', '.join(cls._LEVELS.keys())}")
```
Comment (regisss) [inline optimum/onnxruntime/configuration.py:792]: Nit: I would add at the end of the error message: "got {optimization_level}"
```suggestion
            raise ValueError(f"optimization_level must be in {', '.join(cls._LEVELS.keys())}, got {optimization_level}")
```

File: optimum/onnxruntime/configuration.py [hunk 11]
```diff
@@ -604,73 +608,248 @@ def tensorrt(
...
+
+    _LEVELS = {
+        "O1": {
+            "optimization_level": 1,
+            "enable_transformers_specific_optimizations": False,
+        },
+        "O2": {
+            "optimization_level": 2,
+            "enable_transformers_specific_optimizations": True,
+            "enable_gelu_approximation": True,
```
Comment (regisss) [inline optimum/onnxruntime/configuration.py:761]: Yeah I don't know if `enable_gelu_approximation` makes sense in a medium config :thinking: 
I don't have a strong opinion on this since I don't really know to which extent it impacts the model behaviour.

**Review by regisss — APPROVED** [review]
Great PR @michaelbenayoun :rocket: 
I just left a nit.

File: optimum/onnxruntime/configuration.py (hunk 11 above)
Comment (michaelbenayoun) [reply optimum/onnxruntime/configuration.py]: I just added another optimization level.
